documentation/Using-the-SDK/Advanced/Creating-Your-Own-DataSource.md

```diff
@@ -0,0 +1,204 @@
+    public WebServiceDataSourceAttribute(
+        string serviceUri)
+        : base(typeof(WebServiceDataSource))
+    {
+        this.ServiceUri = new Uri(serviceUri);
+    }
+
+    public Uri ServiceUri { get; }
+}
+````
+
+## Accepts
+
+Finally, by default, all Data Sources of a given type will be routed to 
+Custom Data Sources that are decorated with the corresponding attribute. In
+this case, _ANY_ instance of a WebServiceDataSource will be routed to our Custom
+Data Source, and the IsDataSourceSupported method would have to do interrogation
+of said Data Source to determine if it is supported. Wouldn't it be nice if
+there was a way to say "I only care about certain servces; don't even bother
+with others?" Yes, through the _Accepts_ method. This method allows for you
+to filter incoming DataSource objects before they ever reach your Custom Data
+Source so that your logic is simpler. For example, the FileDataSourceAttribute
+in the SDK uses this method to reject anything that does not match the given
+file extension, so Custom Data Sources have _FileDataSource(".txt")_, for
+example, and only ever see .txt files. You should still implement
+IsDataSourceSupported in your CustomDataSource to do the final determination.
+
+````cs
+// we are decorating Custom Data Source classes, so we must have the correct
+// attribute targets
+[AttributeUsage(AttributeTargets.Class)]
+public sealed class WebServiceDataSourceAttribute
+    : DataSourceAttribute
+{
+    // our constructor will pass the DataSource Type to the base class. We
+    // can take any other parameters that we like to help denote our data. In
+    // this case, we simply say that data on the given server is supported.
+    public WebServiceDataSourceAttribute(
+        string serviceUri)
+        : base(typeof(WebServiceDataSource))
+    {
+        this.ServiceUri = new Uri(serviceUri);
+    }
+
+    public Uri ServiceUri { get; }
+
+    // implement this method to eagerly filter out data sources that are not
+    // accepted by the decorated Custom Data Source.
+    public override bool Accepts(IDataSource dataSource)
+    {
+        // IDataSource will always be of the Type based to the base
+        // class in our constructor.
+        var web = dataSource as WebServiceDataSource;
+
+        // This check is for defensive programming in case dataSource is
+        // ever null.
+        if (web is null)
+        {
+            return false;
+        }
+
+        // We only accept WebServiceDataSources that target our declared service.
+        return this.ServiceUri == web.ServiceUri;
+    } 
+}
+````
+
+# Using Your Data Source 
+
+Now that we have our new Data Source, we will see how to use our new Data Source
+with the Engine.
+
+````cs
+
+namespace Sample
+{
+    [CustomDataSource(
+        /* Guid here */,
+        /* Name here */,
+        /* Description here */)]
+    [WebServiceDataSource("https://www.contoso.com")]
```

Comment:
the https warns of a certificate error before redirecting, maybe put `http` here 😆